src/lexer.cpp

```diff
@@ -412,16 +343,18 @@ TokenArray tokenize(char const* p) {
 		char const* const p0 = p;
 
 		int state = state_count - 1;
-		while (MainLexer::EndStates::Count < state) {
-			state = a.go(state, *p++);
+		int new_state = a.go(state, *p++);
+		while (new_state != MainLexer::EndStates::Error) {
+			state = new_state;
+			new_state = a.go(state, *p++);
 		}
 
```

Comment:
I think I prefer this expressed as an infinite loop with an if-break, to eliminate the need of writing the `new_state` assignment twice.

(if your thing is faster, though, let's use that)

```suggestion
		while(1) {
			int new_state = a.go(state, *p++);
			if (new_state == MainLexer::EndStates::Error) break;
			state = new_state;
		}
```

Reply:
i looked at the asm output, and they both generate the exact same code for the inner loop, but the surrounding code is a bit longer for yours.

There is also this version, which generates a weird instruction I don't know in the inner loop, but otherwise is very similar to my other sugestion

```suggestion
		while (1) {
			int new_state = a.go(state, *p++);
			if (new_state == MainLexer::EndStates::Error) break;
			state = new_state;
		}
```

Reply:
All of these (including yours) have 2 more instructions in the inner loop (for a total of 10) than the original (which has 8)